XenAdmin/SplashScreenContext.cs

```diff
@@ -0,0 +1,120 @@
+﻿/* Copyright (c) Cloud Software Group Holdings, Inc.
+ * All rights reserved. 
+ * 
+ * Redistribution and use in source and binary forms, 
+ * with or without modification, are permitted provided 
+ * that the following conditions are met: 
+ * 
+ * *   Redistributions of source code must retain the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer. 
+ * *   Redistributions in binary form must reproduce the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer in the documentation and/or other 
+ *     materials provided with the distribution. 
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND 
+ * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using System;
+using System.Configuration;
+using System.Reflection;
+using System.Threading;
+using System.Windows.Forms;
+using XenAdmin.Dialogs;
+
+namespace XenAdmin
+{
+    internal class SplashScreenContext : ApplicationContext
+    {
+        private static readonly log4net.ILog _log = log4net.LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);
+
+        private readonly SplashScreen _splashScreen;
+        private MainWindow _mainWindow;
+        private readonly string[] _args;
+
+        public static MainWindow MainWindow { get; private set; }

```

Comment:
I know it's not that great, but I wanted to keep the non-static `_mainWindow` field in place so as to have symmetry with the `_splashscreen` field. An alternative solution is to remove `SplashScreenContext.MainWindow` completely and instead set the `Program.MainWindow` in line 96 - but to do this the latter will have to become `public set` and there is a risk that people may start setting it all over the place. Or we can replace the 492 usages of `Program.MainWindow` with `SplashScreenContext.MainWindow` 😃 
Secretly I'm hoping that one day in the remote future we might be able to get rid of all these static instances because they complicate the logic.